rcljava/src/main/cpp/org_ros2_rcljava_RCLJava.cpp

```diff
@@ -58,20 +59,73 @@ Java_org_ros2_rcljava_RCLJava_nativeCreateNodeHandle(
+  // more infrastructure for that, and we don't expect to call
+  // 'nativeCreateNodeHandle' that often, so I think this is OK for now.
+  jclass java_util_ArrayList =
+    static_cast<jclass>(env->NewGlobalRef(env->FindClass("java/util/ArrayList")));
+  jmethodID java_util_ArrayList_size = env->GetMethodID(java_util_ArrayList, "size", "()I");
+  jmethodID java_util_ArrayList_get = env->GetMethodID(
+    java_util_ArrayList, "get", "(I)Ljava/lang/Object;");
+
+  rcl_arguments_t arguments = rcl_get_zero_initialized_arguments();
+  char ** argv = NULL;
+  jint argc = env->CallIntMethod(cli_args, java_util_ArrayList_size);
+  argv = static_cast<char **>(malloc(argc * sizeof(char *)));
+  for (jint i = 0; i < argc; ++i) {
+    jstring element =
+      static_cast<jstring>(env->CallObjectMethod(cli_args, java_util_ArrayList_get, i));
+    argv[i] = const_cast<char *>(env->GetStringUTFChars(element, nullptr));
+  }
+  rcl_ret_t ret = rcl_parse_arguments(argc, argv, rcl_get_default_allocator(), &arguments);
+  for (jint i = 0; i < argc; ++i) {
+    jstring element =
+      static_cast<jstring>(env->CallObjectMethod(cli_args, java_util_ArrayList_get, i));
+    env->ReleaseStringUTFChars(element, argv[i]);
+  }
+  free(argv);
+  if (ret != RCL_RET_OK) {
+    std::string msg = "Failed to parse node arguments: " + std::string(rcl_get_error_string().str);
+    rcl_reset_error();
+    rcljava_throw_rclexception(env, ret, msg);
+    return 0;
+  }
+
   rcl_node_t * node = static_cast<rcl_node_t *>(malloc(sizeof(rcl_node_t)));
   *node = rcl_get_zero_initialized_node();
 
-  rcl_node_options_t default_options = rcl_node_get_default_options();
-  rcl_ret_t ret = rcl_node_init(
-    node, node_name.c_str(), namespace_.c_str(), context, &default_options);
+  rcl_node_options_t options = rcl_node_get_default_options();
+  options.use_global_arguments = use_global_arguments;
+  options.arguments = arguments;
+  options.enable_rosout = enable_rosout;
+  ret = rcl_node_init(node, node_name.c_str(), namespace_.c_str(), context, &options);
   if (ret != RCL_RET_OK) {
     std::string msg = "Failed to create node: " + std::string(rcl_get_error_string().str);
     rcl_reset_error();
+    free(node);
+    if (rcl_arguments_fini(&arguments) != RCL_RET_OK) {
+      // We are already throwing an exception, just ignore the return here
+    }
     rcljava_throw_rclexception(env, ret, msg);
     return 0;
   }
-  jlong node_handle = reinterpret_cast<jlong>(node);
-  return node_handle;
+
+  ret = rcl_arguments_fini(&arguments);
+  if (ret != RCL_RET_OK) {
+    // We failed to cleanup
+    std::string msg = "Failed to cleanup after creating node: " + std::string(
+      rcl_get_error_string().str);
+    rcl_reset_error();
+    if (rcl_node_fini(node) != RCL_RET_OK) {
+      // We are already throwing an exception, just ignore the return here
+    }
+    free(node);
+    rcljava_throw_rclexception(env, ret, msg);
+    return 0;
+  }
+
+  return reinterpret_cast<jlong>(node);
```

Comment:
meta: the error handling code in this function is quite complicated.

Allocating `node` and `arguments` in a `unique_ptr` with a deleter, and releasing the `unique_ptr` when the function doesn't fail would be a bit clearer.

Reply:
Good idea, let me try that and see how it goes.

Reply:
I ended up doing something different here.

Most of the complexity came from the fact that argument parsing needed a whole new set of code.  To keep the complexity down in `Java_org_ros2_rcljava_RCLJava_nativeCreateNodeHandle`, I ended up moving the argument parsing to its own function.  I think with that in place, the code is now a lot easier to read.

I could *also* do the unique_ptr with release thing, but I don't think it would add too much additional benefit at this point.  But feel free to disagree.

Reply:
> Most of the complexity came from the fact that argument parsing needed a whole new set of code. To keep the complexity down in Java_org_ros2_rcljava_RCLJava_nativeCreateNodeHandle, I ended up moving the argument parsing to its own function. I think with that in place, the code is now a lot easier to read.

:+1: to that.

---

I don't mind strongly, but I think we can deduplicate cleanup code by doing something like this (feel free to ignore the suggestion):

```cpp
std::unique_ptr<rcl_node_t, std::function<void(rcl_node_t *)> node(
  new(std::nothrow) rcl_node_t(),
  [](rcl_node_t * node) {
    if (rcl_node_fini(node) != RCL_RET_OK) {
      // This code is only run when an exception has already been raised
    }
    delete node;
  });
if (!node.get()) {
  return 0;
}
{
  rcl_arguments_t arguments;
  if (!parse_arguments(env, cli_args, &arguments)) {
    // All of the exception setup was done by parse_arguments, just return here.
    return 0;
  }
  rcpputils::scope_exit({
    ret = rcl_arguments_fini(&arguments);
    if (ret != RCL_RET_OK && !ExceptionCheck(env)) {
      // Failed to destroy and an exception has not been raised before, raise one
      std::string msg = "Failed to cleanup after creating node: " + std::string(
        rcl_get_error_string().str);
      rcl_reset_error();
      rcljava_throw_rclexception(env, ret, msg);
    }
  });
  ... // fill node_options
  rcl_ret_t ret = rcl_node_init(node.get(), node_name.c_str(), namespace_.c_str(), context, &options);
  if (ret != RCL_RET_OK) {
    std::string msg = "Failed to create node: " + std::string(rcl_get_error_string().str);
    rcl_reset_error();
    rcljava_throw_rclexception(env, ret, msg);
    return 0;
  }
}
if (ExceptionCheck(env)) {
  return 0;
}
return reinterpret_cast<jlong>(node.release());

Reply:
Thanks, that's a good example.  I think there are 2 problems with that code above:
1.  When the `unique_ptr` to node is first allocated, it is in an undefined state.  That is, we haven't yet run `rcl_node_init`, so all of the fields are essentially UB.  If we fail during `parse_arguments`, for instance, we would then try to run `rcl_node_fini` against the uninitialized `node`, and I think that would lead to UB.
1.  Currently rcljava doesn't depend on `rclcpp`, `osrf_testing_tools_cpp`, or `rcpputils`.  That means that all of the implementations of `scope_exit` that we have aren't currently available.  I definitely don't want to add a dependency on `rclcpp`, so that is out.  We could resolve this either by making a new dependency on `osrf_testing_tools_cpp` or `rcpputils`, or by porting the `scope_exit` stuff into here.  I don't really love either solution, but we could do it.

Neither problem is insurmountable, but I think that together they are enough to make me not want to do this.  But thanks for showing how it could be done, that was really instructive!